### DISCUSSION TITLE
Rails `importmap pin tom-select` pulled wrong version?

### DISCUSSION THREAD

**Discussion by silent-e** [orchidjs/tom-select#830] [Q&A] [answered]
I am using tom-select via importmaps in my Rails app.  I was on v2.3.1 but after running `importmap pin tom-select` to get v2.4.1 it appears to have downloaded the wrong version from the cdn.

I got the following output

```
❯ importmap pin tom-select
Pinning "tom-select" to vendor/javascript/tom-select.js via download from https://ga.jspm.io/npm:tom-select@2.4.1/dist/esm/tom-select.complete.js
Pinning "@orchidjs/sifter" to vendor/javascript/@orchidjs/sifter.js via download from https://ga.jspm.io/npm:@orchidjs/sifter@1.1.0/dist/esm/sifter.js
Pinning "@orchidjs/unicode-variants" to vendor/javascript/@orchidjs/unicode-variants.js via download from https://ga.jspm.io/npm:@orchidjs/unicode-variants@1.1.2/dist/esm/index.js
```

Note the `/esm/` subdirectory it pulled from.  This pulled the module version but I wanted the compiled version from the `/js/` subdirectory on the CDN.

I haven't opened an issue yet but was this version (and other dependencies) incorrectly downloaded because of a change in the `package.json` file?  My javascript-fu is not strong enough to know.

Thoughts?

(e)

**Comment by maxim** [comment, marked as answer]
For anyone coming here, you can get the file like this:

```
bin/importmap pin tom-select/dist/js/tom-select.complete.min.js
mv vendor/javascript/tom-select--dist--js--tom-select.complete.min.js.js vendor/javascript/tom-select.js
```

and then in importmap.rb

```ruby
pin "tom-select" # @2.4.1
```

This seems to bundle tom-select, sifter, unicode-variants and all the plugins. And this line `import TomSelect from "tom-select"` seems to work.

P.S. So far I've discovered everything like the above by trial and error. I [posted some questions](https://discuss.hotwired.dev/t/understanding-bin-importmap-pin-urls/6158) trying to understand the pin urls better, but at the time of this writing haven't gotten any answers.

P.P.S. Every pin in my importmap.rb comes with a big comment above it, explaining all the necessary process to get the dependency correctly, because every single library seems idiosyncratic in the way that it lets you get a fully-bundled esm. I suggest ya'll do the same: big comment about every pin on how you download it, rename it, and pin it.

**Reply by silent-e** [reply]
That's fascinating.  I didn't know you could pin like that.  I've read through your question and ParamagicDev's responses and think I basically understand how that works.  Bit complicated to figure out what particular file you need to reference and then make sure that any updates have to abide by the specific instructions.

Just to add to the knowledge base…. When you run @maxim's `importmap pin` command you will get the following in your `config/importmap.rb` file

```ruby
pin "tom-select/dist/js/tom-select.complete.min.js", to: "tom-select--dist--js--tom-select.complete.min.js.js" # @2.4.2
```

You _could_ just replace the first path with `tom-select` and it would work using the ugly download name.  I like moving it to a nicer filename too.

Could it improve TomSelect to add something like this to their package.json as mentioned by ParamagicDev at https://discuss.hotwired.dev/t/understanding-bin-importmap-pin-urls/6158/4 ?

```javascript
"jsdelivr": "dist/js/tom-select.complete.min.js",
```

Thank you so much @maxim!

**Comment by yskkin** [comment]
https://github.com/rails/importmap-rails/issues/288 seems related

**Comment by gmarziou** [comment]
The version on default cdn jspm is behind jsdelivr: 2.4.3 vs. 2.5.2

So, if you want the latest version you should specify the cdn on command line:

`./bin/importmap pin tom-select --from=jsdelivr`
